src/daft-plan/src/logical_ops/join.rs

```diff
@@ -167,6 +169,37 @@ impl Join {
         }
     }
 
+    fn rename_join_keys(
+        left_exprs: Vec<Arc<Expr>>,
+        right_exprs: Vec<Arc<Expr>>,
+    ) -> (Vec<Arc<Expr>>, Vec<Arc<Expr>>) {
+        left_exprs
+            .into_iter()
+            .enumerate()
+            .map(|(i, left_expr)| {
+                let right_expr = &right_exprs[i];
+
+                // Rename left_expr if it's different from right_expr
+                let renamed_left_expr = if left_expr != *right_expr {
```

Comment:
One more condition I would like to add is that both expressions should be column expressions, since that is the condition for being common join keys. Also, because of that, I would put this after the calls to `resolve_exprs` since that might actually expand some column expressions to getters.

Reply:
For example, we would not want to preserve the naming if both join keys were `col("a.b")`, which would be resolved to something like `col("a").struct.get("b")`

Reply:
This is a relevant point (limit to column expressions), however putting the renaming logic as-is after the `resolve_exprs` would not work. This is because `resolve_exprs` also computes the `left_fields` and `right_fields` which is then iterated over in the next for loop. So the error in this issue still gets thrown when `Schema` is constructed with `on_fields` returned by `resolve_exprs`.

I think this would involve touching more parts of the code. Could you guide me through ?

Reply:
@kevinzwang : could you please confirm my understanding ?

Reply:
Thanks for reminding me! Based on what I see, the fields computed from `resolve_exprs` are only used to determine the output types of the expressions when checking for null type join keys.

What you can do instead of using `left_fields` and `right_fields` is call [.to_field](https://github.com/Eventual-Inc/Daft/blob/main/src/daft-dsl/src/expr.rs#L755) on each of the renamed expression, and check if the dtype of that output is null instead. Do you think you could give that a go?

Reply:
There is a new commit addressing all remarks.